### PR TITLE
Deduplicate field-caps fields in node and ccs responses

### PR DESCRIPTION
**Pull request by dnhatn**
We can reduce the memory usage of field-caps responses from node and remote-cluster responses by sharing field responses with the same name.  I think we can also remove the type deduplicating as we now deduplicate the whole field response.

This change can handle more concurrent field-caps requests targeting many indices.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/fieldcaps/IndexFieldCapabilities.java [hunk 1]
```diff
@@ -62,7 +60,7 @@ public class IndexFieldCapabilities implements Writeable {
     IndexFieldCapabilities(StreamInput in) throws IOException {
         if (in.getVersion().onOrAfter(Version.V_7_7_0)) {
             this.name = in.readString();
-            this.type = typeStringDeduplicator.deduplicate(in.readString());
+            this.type = in.readString();
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/action/fieldcaps/IndexFieldCapabilities.java:63]: I get where this is coming from and like the idea here a lot, but isn't type still going to be a massively repeated string potentially if we have a ton of different fields in a large use-case? Might still be worthwhile to dedup this?

**Review by original-brownbear — COMMENTED** [review]
I like this and it makes sense to me with my understanding of what fields will look like on average, but it's probably best for someone from search to do a real review :)
Thanks Nhat!

File: server/src/main/java/org/elasticsearch/action/fieldcaps/IndexFieldCapabilities.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/fieldcaps/IndexFieldCapabilities.java]: I think we should keep the type deduplicator. I restored it in https://github.com/elastic/elasticsearch/pull/88238/commits/4b9c1aad17746ddfd21192b0d66ce89e3450a70f.

**Comment by dnhatn** [comment]
Thanks Armin. @javanna Would you mind taking a look?

**Comment by javanna** [comment]
hey @dnhatn do you have numbers on the improvement that this brings? I am assuming that this is not needed on 8.x because we already deduplicate based on the mappings hash while in this case we ensure reusing the same instances for fields with same name and definition across multiple indices?

File: server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesFetcher.java [hunk 2]
```diff
@@ -37,9 +37,11 @@
  */
 class FieldCapabilitiesFetcher {
     private final IndicesService indicesService;
+    private final IndexFieldCapabilities.Deduplicator fieldDeduplicator;
 
-    FieldCapabilitiesFetcher(IndicesService indicesService) {
+    FieldCapabilitiesFetcher(IndicesService indicesService, boolean dedupFields) {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesFetcher.java:42]: why do we need to enable deduplication conditionally?

**Comment by dnhatn** [comment]
@javanna Thanks for reviewing. 

> do you have numbers on the improvement that this brings

Without this change, a 3-nodes cluster with an 8G heap couldn't handle 3 concurrent field-caps requests targeting 5000 `metricbeat` indices. With this patch, that cluster handled those field-caps requests without issues. This change also speeds up field-caps requests a bit but not significant.

>  I am assuming that this is not needed on 8.x because we already deduplicate based on the mappings hash while in this case we ensure reusing the same instances for fields with same name and definition across multiple indices

I think we shouldn't push this improvement into 8.x as it won't save much memory and can slow down the serialization.

File: server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesFetcher.java (hunk 2 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesFetcher.java]: I disabled the deduplication for index-level requests from old nodes (i.e., before 7.16) as the deduplication won't help. However, I think we should always enable the deduplication for simplicity. I pushed https://github.com/elastic/elasticsearch/pull/88238/commits/76a0bd169fb488c279482484111e761f57ff3074.

File: server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java [hunk 3]
```diff
@@ -337,6 +337,7 @@ public void messageReceived(FieldCapabilitiesNodeRequest request, TransportChann
                 final Map<String, List<ShardId>> groupedShardIds = request.shardIds()
                     .stream()
                     .collect(Collectors.groupingBy(ShardId::getIndexName));
+                final FieldCapabilitiesFetcher fieldCapabilitiesFetcher = new FieldCapabilitiesFetcher(indicesService);
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java:340]: should we move this back where it used to be?

Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java:340]: We should one Fetcher for each field-caps request as we shouldn't keep the deduplication map after the request completes.

File: server/src/main/java/org/elasticsearch/action/fieldcaps/RequestDispatcher.java [hunk 4]
```diff
@@ -227,7 +227,7 @@ private void sendRequestToNode(String nodeId, List<ShardId> shardIds) {
                     shardRequest,
                     parentTask,
                     TransportRequestOptions.EMPTY,
-                    new ActionListenerResponseHandler<>(listener, FieldCapabilitiesIndexResponse::new)
+                    new ActionListenerResponseHandler<>(listener, is -> new FieldCapabilitiesIndexResponse(is, f -> f))
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/fieldcaps/RequestDispatcher.java]: Yes, I have removed the interface in https://github.com/elastic/elasticsearch/pull/88238/commits/c184ff3726eea59dd154263eca0e29e8ac3e38f3. The impact should be minimal as the no-op deduplication is used only when interacting with old-nodes (before 7.16).

File: server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesIndexResponse.java [hunk 5]
```diff
@@ -34,10 +34,10 @@ public class FieldCapabilitiesIndexResponse extends ActionResponse implements Wr
         this.originVersion = Version.CURRENT;
     }
 
-    FieldCapabilitiesIndexResponse(StreamInput in) throws IOException {
+    FieldCapabilitiesIndexResponse(StreamInput in, IndexFieldCapabilities.Deduplicator fieldDeduplicator) throws IOException {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesIndexResponse.java:37]: Does it need to be provided as a constructor argument? Isn't it always the same type, just created externally? Could we create it within readFields?

**Review by javanna — APPROVED** [review]
left a small comment, LGTM otherwise

File: server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesIndexResponse.java (hunk 5 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesIndexResponse.java:37]: We can't create the deduplicator inside `readFields` as we need to deduplicate fields between index responses. We can avoid passing the deduplicator in the constructor by deferring the deduplication until we have constructed an index response. However, this requires us to iterate and recreate a list of fields.

**Comment by dnhatn** [comment]
Thanks Luca.

File: server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesIndexResponse.java (hunk 5 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/fieldcaps/FieldCapabilitiesIndexResponse.java:37]: thanks for explaining :)

**Pull request merged**